pipeline/src/pipeline.rs

```diff
@@ -27,6 +28,16 @@ impl Pipeline {
     }
 }
 
+// Manual implemention is needed here as we want to only serialize as series of nodes.
+impl Serialize for Pipeline {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.nodes().serialize(serializer)
```

Comment:
This is kind of wasteful (as is `nodes` returning a `Vec` rather than an iterator`) but I guess it's fine since we will replacing this soon.

Reply:
"as is" -> "as in"? Is there a way to transform the iterator returned by `HashMap::values()` to something serializable?

Reply:
other than something requiring allocation (like `Vec`) that is.

Reply:
> "as is" -> "as in"?

No. To reword: This is kind of wasteful (and `nodes` returning a `Vec` rather than an iterator is kind of wasteful too) …

> Is there a way to transform the iterator returned by `HashMap::values()` to something serializable?

No need to transform it, just serialize the elements in a loop using [SerializeSeq](https://docs.rs/serde/1.0.114/serde/ser/trait.SerializeSeq.html).

Reply:
>>     "as is" -> "as in"?
> 
> No. To reword: This is kind of wasteful (and nodes returning a Vec rather than an iterator is kind of wasteful too) …

Ah ok. I've changed to using `SerializeSeq` now.

Regarding `nodes()` returning a `Vec`, I made that choice because I didn't want to expose the internal detail of `HashMap` being used and `HashMap::values` gives us a `std::collections::hash_map::Values`. It wasn't very obvious to me, how to transform that into something that doesn't expose map details either so i went for a bit of wasteful but simpler approach.

Reply:
> It wasn't very obvious to me, how to transform that into something that doesn't expose map details either

What you're looking for is `-> impl Iterator<Item = &Node>`. Although it may be that the borrow checker isn't happy with that and it would have to be

```rust
pub fn nodes<'a>(&'a self) -> impl Iterator<Item = &'a Node> {
```

or even

```rust
pub fn nodes<'a>(&'a self) -> impl Iterator<Item = &'a Node> + 'a {
```

Reply:
both of those will be fine and seems that's exactly what [I get](https://doc.rust-lang.org/std/collections/hash_map/struct.Values.html#impl-Iterator). I just had to click the expand button there but that's what i meant by "it wasn't obvious to me". :) I'll try and fix it as well.. Thanks!

Reply:
In the end it was super easy. :) PTALA. :)